File: server/schema/schema.js
// define the graph schema

const graphql = require('graphql');
const { 
	GraphQLObjectType, 
	GraphQLString, 
	GraphQLSchema,
	GraphQLInt,
	GraphQLID, // id will allow int and string
	GraphQLList,
	GraphQLNonNull
	} 

= graphql;

const _ = require('lodash');
const Book = require('../models/book')
const Author = require('../models/author')

//var books = [
//	{name: "book a", id: '1', authorid:'1'},
//	{name: "book b", id: '2', authorid:'2'},
//	{name: "book c", id: '3', authorid:'3'},
//	{name: "book d", id: '4', authorid:'1'},
//	{name: "book e", id: '5', authorid:'2'},
//	{name: "book f", id: '6', authorid:'3'},
//]
//
//var authors = [
//	{name:'Hsin', age:30, id:'1'},
//	{name:'Jess', age:30, id:'2'},
//	{name:'Kyle', age:6, id:'3'}
//]

// instantiate object to define a Book type
const BookType = new GraphQLObjectType({
	name: 'Book',

	// notice here we use a function for "fields" rather than an object. this is because we want to treat it as a function so AuthorType can be used before it's defined (we have AuthorType defined afterwards. When we use a function this won't be concerning since a function will only be executed when it's needed.
	fields: () => ({
		id: {
			type: GraphQLID
		},
		name: {
			type: GraphQLString
		},
		genre: {
			type: GraphQLString
		},
		// here comes the relationship
		author: {
			type: AuthorType,
			resolve(parent, args){ 
			    // parent will return the query result
				// where you can take the info and do things subsquently     
				//console.log(parent)
				return Author.findById(parent.authorId)
		    	//return _.find(authors,{id:parent.authorid});	
			}
		}
	})
});


const AuthorType = new GraphQLObjectType({
	name: 'Author',
	fields: () => ({
		id: {
			type: GraphQLID
		},
		name: {
			type: GraphQLString
		},
		age: {
			type: GraphQLInt
		},
		books: {
			type: new GraphQLList(BookType),
			resolve(parent, args){
				console.log(parent)
				return Book.find({authorId: parent.id});
			    //return	_.filter(books, {authorid:parent.id}); // filter to get back filtered array; _.find is for returning one found result
			}
		}
	})
});

// define the root query type
// this is different than the BookType since we will define every different type of rootquery
const RootQuery = new GraphQLObjectType({
	name: 'RootQueryType',
	// the fields are not function
	fields: {
		// here we just use objects rather than functions like BookType because BookType and AuthorType has been defined.
		book: {
			type: BookType,
			args: {id:{type:GraphQLID}},
			resolve(parent, args){
			// this gets run when client sends a request
			//code to get data from db
			// use lodash here
			//return _.find(books, {id:args.id});
			return Book.findById(args.id)
			}
		},
		author: {
			type: AuthorType,
			args: {id:{type:GraphQLID}},
			resolve(parent, args){
			// this gets run when client sends a request
			//code to get data from db
			// use lodash here
			//return _.find(authors, {id:args.id});
			return Author.findById(args.id)
			}
		},
		books: {
			type: new GraphQLList(BookType),
//			args: {
//				author:{
//					fields:{
//						id: {
//							type: GraphQLID
//						},
//						name: {},
//						age: {}
//					}
//				},
//			},
			resolve(parent, args){
				return Book.find({}) // will return all records
				//return books;
			}
		},
		authors: {
			type: new GraphQLList(AuthorType),
			resolve(parent, args){
				return Author.find({})
				//return authors;
			}
		}
	}
});

const Mutation = new GraphQLObjectType({
	name: 'Mutation',
	fields: {
		addAuthor: {
			type: AuthorType,
			args: {
				name: {type: new GraphQLNonNull(GraphQLString)},
				age: {type: new GraphQLNonNull(GraphQLInt)}
			},
			resolve(parent, args){
				let author = new Author({
					name: args.name,
					age: args.age
				});
				return author.save(); // save this to the database
			}
		},
		addBook: {
			type: BookType,
			args: {
				name: {type: new GraphQLNonNull(GraphQLString)},
				genre: {type: new GraphQLNonNull(GraphQLString)},
				authorId: {type: new GraphQLNonNull(GraphQLID)}
			},
			resolve(parent, args){
				let book = new Book({
					name: args.name,
					genre: args.genre,
					authorId: args.authorId
				})
				return book.save();
			}

		}
	}
})

module.exports = new GraphQLSchema({
    // define the queries that users are allowed to use
	query: RootQuery,
	mutation: Mutation
});

File: python_example/schema.py
import graphene


books = [
	{'book_id':"1", 'book_name': 'book1', 'genre': 'fiction', 'author_id': '1'},
	{'book_id':"2", 'book_name': 'book2', 'genre': 'history', 'author_id': '2'},
	{'book_id':"3", 'book_name': 'book3', 'genre': 'fiction', 'author_id': '1'},

]


authors = [
	{'author_id': "1", 'author_name': 'Hsin', 'age': 40},
	{'author_id': "2", 'author_name': 'Kyle', 'age': 6},
]


class BookType(graphene.ObjectType):
	book_id = graphene.ID(description="book id")
	book_name = graphene.String(name="bookName", default_value="unknown")
	genre = graphene.String()
	author = graphene.Field(lambda : AuthorType)

	def resolve_author(parent, info):
		author_id = parent['author_id']
		for author in authors:
			if author['author_id'] == author_id:
				return author


class AuthorType(graphene.ObjectType):
	author_id = graphene.ID()
	author_name = graphene.String(name="authorName", default_value="unknown")
	age = graphene.Int()
	books = graphene.List(BookType)
	
	def resolve_books(parent, info):
		author_id = parent['author_id']
		return [b for b in books if b['author_id']==author_id]


class BookInput(graphene.InputObjectType):
		book_id = graphene.ID()
		book_name = graphene.String()
		genre = graphene.String()
		author_id = graphene.ID()


class AddBook(graphene.Mutation):
	# define the arguments you need when adding a book (mutation)
	class Arguments:
		book_data = BookInput(required=True)
	
	# fields for AddBook
	ok = graphene.Boolean()
	book = graphene.Field(BookType)
    
	@staticmethod
	def mutate(root, info, book_data=None):
		ok = True
		# you can create a BookType Object like this
		book = BookType(book_id=book_data.book_id, book_name=book_data.book_name, genre=book_data.genre)
		# update db
		books.append({'book_id': book_data.book_id, 'book_name': book_data.book_name, 'genre': book_data.genre, 'author_id': book_data.author_id})
		
		# you can specify the return value
		return AddBook(ok=ok, book=book)


class Mutation(graphene.ObjectType):
	# define fields for Mutation
	add_book = AddBook.Field()


class RootQuery(graphene.ObjectType):
	book = graphene.Field(BookType, book_id=graphene.ID(required=True))
	
	author = graphene.Field(AuthorType, author_id=graphene.ID(required=True))

	def resolve_book(parent, info, book_id):
		for book in books:
			if book['book_id'] == book_id:
				return book

	def resolve_author(parent, info, author_id):
		for author in authors:
			if author['author_id'] == author_id:
				return author

if __name__ == '__main__':

	schema = graphene.Schema(query=RootQuery, mutation=Mutation)

	book_query = '''
    	query book {
			book(bookId: "1") {
				bookId
				bookName
				genre
				author {
					authorName
					age
				}
			}
		
		}
	'''
	
	author_query = '''
		query author {
			author(authorId: "1") {
				age
				authorId
				authorName
				books {
					bookName
					genre
					bookId
				}
			}
		}

	'''
	
	comb_query = '''
		query comb {
			book(bookId:2){
				bookId
				bookName
				genre
			}
			author(authorId:1){
				authorId
				authorName
				age
			}
		}

	'''

	add_book_mutation = '''
		mutation addBookMutation {
			addBook(bookData: {bookId: "4", bookName: "book 4", genre: "science", authorId: "4"}
			) {
				book {
					bookName,
					bookId
				}
			}
		}
	'''

	result = schema.execute(add_book_mutation)
	print(result.data)
	print(books)
